packages/core/lib/scope.js

```diff
@@ -89,9 +89,16 @@ const resolveVarScope = (name, scope) => {
  * @param {Scope} scope The scope to use
  * @param {any => boolean} predicate An optional predicate with which to test the resolved value
  */
-const resolveInternal = (name, scope, predicate = () => true) => {
+const resolveInternal = (
```

Comment:
No, I specifically chose for all these functions to always throw an error. Instead of telling it that you don't want an error, just wrap it in a `try ... catch`. 

Don't add an additional option just for this

Reply:
Retuning any value, like  `false`, is misleading, because you don't know whether the value of `internal[name]` is `false`, or if it simply did not find a value.

throwing an error is unambiguous

Reply:
Passing around errors that aren't necessarily errors is bad practice IMO for a couple of reasons. I don't like it mostly because it adds 5-10 lines of unnecessary code anywhere where an error from this function isn't actually an error. (setting up the try catch, let the variables outside, match to make sure it's the right error) which makes the code ugly. Rather than return false I would prefer to define another function which checks for existence only and returns a boolean given the issues you mentioned with trying to overload this one.

Reply:
> Passing around errors that aren't necessarily errors is bad practice IMO for a couple of reasons.

This is really nuanced/opinionated. As in, I've had quite a few discussions with people about this, and infact, I'm pretty sure the MS interviewer hated me because I chose Errors over returning a `null` or something for an interivew problem.

Anyway, my consensus over the years has been - 1) consistency. i.e. pick either one and stick to it. And 2) Prefer Errors in JS. The reason for this is, languages where people tend to return instead of throw, usually have a better syntax to create Maybe monads (Rust, Haskell and OCaml are great examples). In JS, if we wanted to start support a return value instead of Errors, while staying consistent (see point 1), then we need to have a "Maybe" monad type/class/fn somewhere. Effectively, instead of retuning `value` or `false`|`null`, we'd need to return `{ success: true, value: ... }` and `{ success: false }`. This is actually exactly how Rust implements Maybe monads

See this post. It's a great call for Maybe Monads: https://hoverbear.org/blog/option-monads-in-rust/#not-just-a-null

JS is *not* a great language for this. Error throwing/catching is the convention for "does not exist", apart from `undefined` which I personally dislike when the a valid value is `null` (just makes things really confusing for the reader.

And, instead of creating a bunch of utilites to pass around and manage Maybe monads in Javascript, it's far easier to just create a HOF 

```javascript
const throws = (fn) => (...args) => try { return { success: true, value:  fn(...args) } } catch (e) { return { success: false } }
```

this lets you convert any error throwing function into a maybe monad. Just one function. nice and eay

>  don't like it mostly because it adds 5-10 lines of unnecessary code anywhere where an error from this function isn't actually an error. (setting up the try catch, let the variables outside, match to make sure it's the right error) which makes the code ugly.

the `throws` function above solves all these issues

> Rather than return false I would prefer to define another function which checks for existence only and returns a boolean given the issues you mentioned with trying to overload this one.

Yup overloading is bad, and while creating an addition function is viable. (i.e. `resolveInternal` and `strictResolveInteral` or something). I think you'd agree that a HOF like `throws` is far more elegant? 

The beauty of functional programming over overloading (which is really an OOP context)

Reply:
Also, on a different note

I hate `function x(name, scope, predicate = () => true, someOpt = true, anotherOpt = false)`

I mean, I've done that in a lot places. but it makes the code really ugly. Infact, even with predicate, I think I should've moved all the args into a single argument object. It makes the code far more readable:

`x("some", scope, x => !!x.y.z, false, true)` vs
```javascript
x({
  name: "some",
  scope,
  predicate: x => !!x.y.z,
  someOpt: false,
  anotherOpt: true
})
```